### PR TITLE
IPv6 dual stack support for AKO

### PR DESCRIPTION
**Pull request by aaha97**

### CODE REVIEW

File: internal/lib/lib.go [hunk 1]
```diff
@@ -1725,3 +1727,18 @@ func GetThrottle(key string) *int32 {
...
+	v6mask, _ := strconv.Atoi(strings.Split(vipNetwork.V6Cidr, "/")[1])
+
+	v6cidr := avimodels.IPAddrPrefix{
+		IPAddr: &avimodels.IPAddr{
+			Addr: &v6addr,
+			Type: proto.String("V6"),
+		},
+		Mask: proto.Int32(int32(v6mask)),
+	}
+	vip.AutoAllocateIPType = proto.String("V4_V6")
```
Comment (akshayhavile) [inline internal/lib/lib.go:1742]: Here We are assuming both `v4` and `v6` has been provided. Can user just specify `v6`?

Comment (akshayhavile) [inline internal/lib/lib.go:1742]: `AutoAllocateIPType`  has some license based restrictions. Can you please check that?

File: internal/rest/avi_obj_vsvip.go [hunk 2]
```diff
@@ -646,6 +665,9 @@ func networkNamesToVips(vipNetworks []akov1alpha1.AviInfraSettingVipNetwork, ena
 		}
 		newVip.SubnetUUID = proto.String(vipNetwork.NetworkName)
 		vipList = append(vipList, newVip)
+		if vipNetwork.V6Cidr != "" {
```
Comment (akshayhavile) [inline internal/rest/avi_obj_vsvip.go:668]: Did you test in cloud env? Can you please update your testing?

File: internal/rest/avi_obj_vsvip.go [hunk 3]
```diff
@@ -201,6 +204,9 @@ func (rest *RestOperations) AviVsVipBuild(vsvip_meta *nodes.AviVSVIPNode, vsCach
 						}
 					}
 				}
+				if vipNetwork.V6Cidr != "" {
```
Comment (akshayhavile) [inline internal/rest/avi_obj_vsvip.go:207]: Line number 187, 188: will be true if only v6 given. I think we should not print warning if v6cidr is given.

File: helm/ako/values.yaml [hunk 4]
```diff
@@ -50,6 +50,7 @@ NetworkSettings:
   # vipNetworkList:
   #  - networkName: net1
   #    cidr: 100.1.1.0/24
+  #    v6cidr: fdXY:IJKL:MNOP:15::/64
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply helm/ako/values.yaml:53]: Also, please add a valid ipv6 CIDR as an example.

File: internal/lib/lib.go [hunk 5]
```diff
@@ -34,11 +34,13 @@ import (
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/clients"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/session"
+	"google.golang.org/protobuf/proto"
 
 	"github.com/Masterminds/semver"
 	routev1 "github.com/openshift/api/route/v1"
 	oshiftclient "github.com/openshift/client-go/route/clientset/versioned"
 	"github.com/vmware/alb-sdk/go/models"
+	avimodels "github.com/vmware/alb-sdk/go/models"
```
[the remark this answers is not in the document]
Comment (swathinsankaran) [reply internal/lib/lib.go:43]: is this import the same as the import in L42?

File: helm/ako/crds/ako.vmware.com_aviinfrasettings.yaml [hunk 6]
```diff
@@ -29,6 +29,8 @@ spec:
                           type: string
                         cidr:
                           type: string
+                        v6cidr:
```
Comment (swathinsankaran) [inline helm/ako/crds/ako.vmware.com_aviinfrasettings.yaml:32]: please run `make sync-crd-files` to sync the crd changes with ako-operator

File: internal/rest/avi_obj_vsvip.go [hunk 7]
```diff
@@ -487,8 +502,14 @@ func (rest *RestOperations) AviVsVipCacheAdd(rest_op *utils.RestOp, vsKey avicac
 						utils.AviLog.Infof("key: %s, msg: invalid type for vip in vsvip: %s", key, name)
 						continue
 					}
+					ipType := "V4_ONLY"
+					V4, V6, V4_V6 := "V4_ONLY", "V6_ONLY", "V4_V6"
```
Comment (swathinsankaran) [inline internal/rest/avi_obj_vsvip.go:506]: please add this as constants.

File: internal/lib/lib.go [hunk 8]
```diff
@@ -1769,3 +1770,22 @@ func GetThrottle(key string) *int32 {
 	throttle := int32(throttle[key])
 	return &throttle
 }
+
+func UpdateV6(vip *models.Vip, vipNetwork *akov1alpha1.AviInfraSettingVipNetwork) {
+	v6addr := strings.Split(vipNetwork.V6Cidr, "/")[0]
+	v6mask, _ := strconv.Atoi(strings.Split(vipNetwork.V6Cidr, "/")[1])
```
Comment (swathinsankaran) [inline internal/lib/lib.go:1776]: please avoid calling strings.Split two times here.

**Review by swathinsankaran — COMMENTED** [review]
LGTM.
some minor comments and also please update the scenarios you have tested so far.

**Comment by swathinsankaran** [comment]
build ako

**Comment by swathinsankaran** [comment]
build ako

**Comment by apalsule** [comment]
@aaha97 Please update the results of all the FT runs on the PR.

**Comment by swathinsankaran** [comment]
build ako

File: internal/lib/lib.go (hunk 1 above)
Comment (aaha97) [reply internal/lib/lib.go]: updated to allow v6 only

Comment (aaha97) [reply internal/lib/lib.go]: this will be documented

File: internal/rest/avi_obj_vsvip.go (hunk 2 above)
Comment (aaha97) [reply internal/rest/avi_obj_vsvip.go:689]: only vCenter cloud is supported

File: internal/rest/avi_obj_vsvip.go (hunk 7 above)
Comment (aaha97) [reply internal/rest/avi_obj_vsvip.go]: fixed

File: internal/lib/lib.go (hunk 8 above)
Comment (aaha97) [reply internal/lib/lib.go]: fixed

**Comment by aaha97** [comment]
build ako

File: internal/lib/lib.go [hunk 9]
```diff
@@ -34,6 +34,7 @@ import (
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/pkg/utils"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/clients"
 	"github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/third_party/github.com/vmware/alb-sdk/go/session"
+	"google.golang.org/protobuf/proto"
```
[the remark this answers is not in the document]
Comment (aaha97) [reply internal/lib/lib.go]: done

File: helm/ako/crds/ako.vmware.com_aviinfrasettings.yaml (hunk 6 above)
Comment (aaha97) [reply helm/ako/crds/ako.vmware.com_aviinfrasettings.yaml:32]: done

**Review by swathinsankaran — APPROVED** [review]

**Comment by swathinsankaran** [comment]
build ako

File: internal/nodes/avi_vrf_translator.go [hunk 10]
```diff
@@ -89,6 +93,20 @@ func (o *AviObjectGraph) addRouteForNode(node *v1.Node, vrfName string, routeid
 			break
 		}
 	}
+	if ipFamily == v6Type {
+		if lib.GetCNIPlugin() == lib.CALICO_CNI {
+			if ip, ok := node.Annotations["projectcalico.org/IPv6Address"]; ok {
+				nodeIP6 = strings.Split(ip, "/")[0]
+			}
+			if nodeIP6 == "" {
+				utils.AviLog.Errorf("Error in fetching nodeIPv6 for %v", node.ObjectMeta.Name)
```
Comment (akshayhavile) [inline internal/nodes/avi_vrf_translator.go:102]: Why are we erroring out here?  Do we want to have route table have only ipv6 route        
 -->next hop IPV6 and IPV4 route --> next hop ipv4

File: internal/rest/avi_obj_vsvip.go [hunk 11]
```diff
@@ -487,8 +502,13 @@ func (rest *RestOperations) AviVsVipCacheAdd(rest_op *utils.RestOp, vsKey avicac
 						utils.AviLog.Infof("key: %s, msg: invalid type for vip in vsvip: %s", key, name)
 						continue
 					}
+					ipType := "V4_ONLY"
```
Comment (akshayhavile) [inline internal/rest/avi_obj_vsvip.go:505]: we can use already defined constant here.

File: internal/nodes/avi_vrf_translator.go (hunk 10 above)
Comment (aaha97) [reply internal/nodes/avi_vrf_translator.go:102]: yes, that is the expectation

File: internal/lib/lib.go [hunk 12]
```diff
@@ -1769,3 +1771,45 @@ func GetThrottle(key string) *int32 {
...
+
+func UpdateV6(vip *models.Vip, vipNetwork *akov1alpha1.AviInfraSettingVipNetwork) {
+	if vipNetwork.Cidr != "" {
+		vip.AutoAllocateIPType = proto.String("V4_V6")
+	} else {
+		vip.AutoAllocateIPType = proto.String("V6_ONLY")
+	}
+}
+
+func GetIPFamily() string {
```
[the remark this answers is not in the document]
Comment (aaha97) [reply internal/lib/lib.go]: fixed, setting the IPFamily during InitializeAKC and retrieving it later

File: internal/nodes/avi_model_nodes.go [hunk 13]
```diff
@@ -1196,7 +1196,7 @@ func (v *AviVSVIPNode) CalculateCheckSum() {
 	if len(v.VipNetworks) > 0 {
 		var vipNetworkStringList []string
 		for _, vipNetwork := range v.VipNetworks {
-			vipNetworkStringList = append(vipNetworkStringList, vipNetwork.NetworkName+":"+vipNetwork.Cidr)
+			vipNetworkStringList = append(vipNetworkStringList, vipNetwork.NetworkName+":"+vipNetwork.Cidr+":"+vipNetwork.V6Cidr)
```
[the remark this answers is not in the document]
Comment (aaha97) [reply internal/nodes/avi_model_nodes.go]: updated the checksum code

File: internal/rest/avi_obj_vsvip.go (hunk 11 above)
Comment (aaha97) [reply internal/rest/avi_obj_vsvip.go]: fixed

**Review by akshayhavile — APPROVED** [review]
LGTM

**Comment by akshayhavile** [comment]
build ako

**Comment by aaha97** [comment]
adding UT results 

`ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/servicesapitests	22.288s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/oshiftroutetests	792.290s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/namespacesynctests	125.684s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multiclusteringresstests	76.758s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/bootuptests	1.907s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/dedicatedvstests	79.628s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/npltests	76.185s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/advl4tests	13.523s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/infratests	61.728s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/ingresstests	(cached)

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/integrationtest	87.536s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/k8stest	32.111s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/evhtests	280.598s

ok  	github.com/vmware/load-balancer-and-ingress-services-for-kubernetes/tests/multicloudtests	170.926s`

**Review by swathinsankaran — APPROVED** [review]

**Comment by swathinsankaran** [comment]
@akshayhavile could you please approve and merge this PR? Arif has attached the UT results.

**Review by akshayhavile — APPROVED** [review]

**Comment by akshayhavile** [comment]
@apalsule : UTs are passing locally so force merging PR.

**Pull request merged**